stl/inc/memory

```diff
@@ -3195,34 +3195,34 @@ _CXX20_DEPRECATE_OLD_SHARED_PTR_ATOMIC_SUPPORT bool atomic_compare_exchange_stro
 
 #if _HAS_CXX20
 template <class _Ty>
-struct alignas(2 * sizeof(void*)) _Atomic_ptr_base {
+class alignas(2 * sizeof(void*)) _Atomic_ptr_base {
     // overalignment is to allow potential future use of cmpxchg16b
 protected:
     constexpr _Atomic_ptr_base() noexcept = default;
 
     _Atomic_ptr_base(_Ty* const _Px, _Ref_count_base* const _Ref) noexcept
-        : _Ptr(_Px), _Repptr(reinterpret_cast<const uintptr_t>(_Ref)) {}
+        : _Ptr(_Px), _Repptr(reinterpret_cast<uintptr_t>(_Ref)) {}
 
-    _Ref_count_base* _Lock_and_load() const noexcept {
-        static constexpr uintptr_t _Low_bit = static_cast<uintptr_t>(1u);
-        uintptr_t _Rep                      = _Repptr & ~_Low_bit;
-        while (!_Repptr.compare_exchange_strong(_Rep, _Rep | _Low_bit)) {
+    _NODISCARD _Ref_count_base* _Lock_and_load() const noexcept {
+        constexpr uintptr_t _Low_bit = 1;
+        uintptr_t _Rep               = _Repptr.load(memory_order::acquire) & ~_Low_bit;
+        while (!_Repptr.compare_exchange_weak(_Rep, _Rep | _Low_bit, memory_order::acquire)) {

```

Comment:
I am not an expert here, but should `compare_exchange_weak` use `memory_order::acq_rel` so that when it loads (due to failure) we acquire and when it stores (due to success) we release?

Reply:
I don't think we need to prevent subsequent ops from entering the critical region -- this is a lock acquisition -- so acquire should be OK as far as I know.

Reply:
Is there a potential issue if you have two consecutive calls to `atomic<shared_ptr<_Ty>>::store()` (or other operations).
```
atomic<shared_ptr<_Ty>> p1;
atomic<shared_ptr<_Ty>> p2;

// Thread 1
p1.store(make_shared<_Ty>());
p2.store(make_shared<_Ty>());

// Thread 2
p2.store(make_shared<_Ty>());
p1.store(make_shared<_Ty>());
```

Each thread would expand to execute roughly the following:
```
{
  auto tmp = make_shared<_Ty>();
  auto _Rep = p1._Lock_and_load(); // acquire semantics
  swap(p1._Ptr, tmp._Ptr);
  p1._Store_and_unlock(_Rep); // release semantics
  tmp._Rep = _Rep;
}
{
  auto tmp = make_shared<_Ty>();
  auto _Rep = p2._Lock_and_load(); // acquire semantics
  swap(p2._Ptr, tmp._Ptr);
  p2._Store_and_unlock(_Rep); // release semantics
  tmp._Rep = _Rep;
}
```

But there's nothing here that prevents the compiler from rearranging the `p2._Lock_and_load()` call from the second op to be reordered before the `p1._Store_and_unlock()` (acquire-op can be reordered before a prior release-op).
Thus thread 1 might end up holding on to both locks at once.

If the same transformation happened for Thread 2 then you could end up with a deadlock scenario.
- Thread 1 acquired p1 lock and was trying to acquire p2 lock before releasing p1 lock
- Thread 2 acquired p2 lock and was trying to acquire p1 lock before releasing p2 lock

So I think there needs to be stronger semantics on at least one of the operations (lock or unlock) to prevent this kind of reordering with operations on other atomic vars.

Reply:
@lewissbaker  Thank you!

Reply:
Also, I don't think this would honor the `memory_order` parameter passed to `load()` or `store()` if the user passes `seq_cst`.

I think you would need to use `memory_order::seq_cst` for the `_Repptr.compare_exchange_weak()` call in `_Lock_and_load()` and (I think) also for the `_Repptr.store()` call in `_Store_and_release()` to give the correct semantics to `atomic<shared_ptr<T>>::load()/store()` when passed `memory_order::seq_cst`.